Update the Canvas2d Layer proposal to support filters. (#30)
fserb/canvas2D: spec/layers.md

### Unmatched calls
An `endLayer()` without a `beginLayer()` is considered malformed and throws an exception. See [here for an analysis of the alternatives considered](https://docs.google.com/document/d/1jeLn8TbCYVuFA9soUGTJnRjFqLkqDmhJElmdW3w_O4Q/edit#heading=h.dilmo33w0023).

### Interation with `save()`/`restore()`
To be consistent with how [unmatched calls](#unmatched-calls) are handled, invalid mixes of `save()`/`restore()` and `beginLayer()`/`endLayer()` calls are considered malformed and throws an exception. For instance, the following are all considered malformed and throw exceptions:

```js
save(); endLayer();  // No matching beginLayer() in current save() level.
beginLayer(); save(); endLayer();  // No matching beginLayer() in current save() level.
save(); beginLayer(); restore();  // No matching save() in current layer.
beginLayer(); restore();  // No matching save() in current layer.
```

### Call to `ctx.reset()` inside layer

> [By design](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-reset), calling `ctx.reset()` resets the whole canvas and bring it back to it's original default state. This includes clearing the context's drawing state stack. Therefore, calling `ctx.reset()` after calls to `ctx.beginLayer()` must discard all pending layers. As described in the [Unclosed Layers section](#unclosed-layers), doing `ctx.beginLayer(); ctx.reset(); ctx.endLayer();` is malformed and throws an exception since `endLayer()` has no matching `beginLayer()` (it was discarded by `reset()`). 

Reviewer: ```suggestion
[By design](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-reset), calling `ctx.reset()` resets the whole canvas and brings it back to its original default state. This includes clearing the context's drawing state stack. Therefore, calling `ctx.reset()` after calls to `ctx.beginLayer()` must discard all pending layers. As described in the [Unclosed Layers section](#unclosed-layers), doing `ctx.beginLayer(); ctx.reset(); ctx.endLayer();` is malformed and throws an exception since `endLayer()` has no matching `beginLayer()` (it was discarded by `reset()`). 
```

### Call to `ctx.clearRect()` inside layer
[By design](https://html.spec.whatwg.org/multipage/canvas.html#drawing-rectangles-to-the-bitmap) `ctx.clearRect` behaves like `crx.fillRect`, with the difference being that it paints with transparent black instead of the current fill style. Therefore, calling `ctx.clearRect()` inside a layer writes transparent black to the pixels in that layer, not directly in the parent canvas or layer.

### Reading canvas content with unclosed layers

> APIs like `ctx.getImageData()` or `ctx.drawImage(canvas)` effectively renders a canvas frame. As described in the [Unclosed layers section](#unclosed-layers), any frame rendered while layers are active has the same effect as closing the layers, calling `ctx.getImageData()` and re-opening all the layers as if they were never closed.

Reviewer: ```suggestion
APIs like `ctx.getImageData()` or `ctx.drawImage(canvas)` effectively render a canvas frame. As described in the [Unclosed layers section](#unclosed-layers), any frame rendered while layers are active has the same effect as closing the layers, calling `ctx.getImageData()`, and re-opening all the layers as if they were never closed.
```

### Call to `ctx.putImageData()` inside layer
[By design](https://html.spec.whatwg.org/multipage/canvas.html#pixel-manipulation), `putImageData()` writes pixels to the canvas wholesale, bypassing globalAlpha, shadow attributes and globalCompositeOperation. To be consistent with this, `putImageData()` must also bypass layers and write directly to the canvas underneath, or else, the pixel written would be affected by the layer's filter, blending or compositing.

### Call to `ctx.drawImage()` inside layer

> [By design](https://html.spec.whatwg.org/multipage/canvas.html#drawing-images), `drawImage()` is affected by globalAlpha, attributes and globalCompositeOperation. To be consistent, calling `drawImage()` inside a layer write the image to that layer, which will in turn be filtered/blended/composited to the parent.

Reviewer: ```suggestion
[By design](https://html.spec.whatwg.org/multipage/canvas.html#drawing-images), `drawImage()` is affected by globalAlpha, attributes and globalCompositeOperation. To be consistent, calling `drawImage()` inside a layer writes the image to that layer, which will in turn be filtered/blended/composited to the parent.
```

```js
  ctx.beginPath();
  ctx.rect(40, 40, 75, 50);
  ctx.beginLayer('drop-shadow(5px 5px 3px black)');
  ctx.stroke();
  ctx.endLayer();
```

### Interaction with `ctx.filter = ...;`
Some Canvas2D implementations (Chrome and Firefox) shipped a feature where filters can be specified on the context directly, by doing `ctx.filter = <some filter>`. For those implementations, calling `ctx.beginLayer(my_filter); ctx.endLayer()` is equivalent to:
```
  [original_filter, ctx.filter] = [ctx.filter, my_filter];
  beginLayer();
  endLayer();
  ctx.filter = original_filter;
```

Specifying a filter on both the context and the layer nests the two filters one into another. For instance,
```
ctx.filter = 'contrast(175%)';
ctx.beginLayer('brightness(3%)')
```
is the same as:
```
ctx.beginLayer('brightness(3%) contrast(175%)');
```

See [here for an analysis of the alternatives considered](https://docs.google.com/document/d/1jeLn8TbCYVuFA9soUGTJnRjFqLkqDmhJElmdW3w_O4Q/edit#heading=h.6u3unyd2kkpo).

## Alternatives considered

A full analysis of the considered alternatives can be found in [this document](https://docs.google.com/document/d/1jeLn8TbCYVuFA9soUGTJnRjFqLkqDmhJElmdW3w_O4Q/edit#)

> In particual:

Reviewer: ```suggestion
In particular:
```

- An analysis of the different API styles considered is done in the [Layer creation API](https://docs.google.com/document/d/1jeLn8TbCYVuFA9soUGTJnRjFqLkqDmhJElmdW3w_O4Q/edit#heading=h.35wlbbo9qx59) section (for instance, `beginLayer(); endLayer();` vs. `layer=beginLayer(); layer.draw();`).

> - Different strategies for specifying layer rendering attribues were compared in the [Layer parameters](https://docs.google.com/document/d/1jeLn8TbCYVuFA9soUGTJnRjFqLkqDmhJElmdW3w_O4Q/edit#heading=h.a7g8s991icf1) section (for instance: `ctx.globalAlpha=0.5; ctx.beginLayer('blur(4px)');` vs `ctx.beginLayer({filter: 'blur(4px)', alpha: 0.5"});`).

Reviewer: ```suggestion
- Different strategies for specifying layer rendering attributes were compared in the [Layer parameters](https://docs.google.com/document/d/1jeLn8TbCYVuFA9soUGTJnRjFqLkqDmhJElmdW3w_O4Q/edit#heading=h.a7g8s991icf1) section (for instance: `ctx.globalAlpha=0.5; ctx.beginLayer('blur(4px)');` vs `ctx.beginLayer({filter: 'blur(4px)', alpha: 0.5"});`).
```
